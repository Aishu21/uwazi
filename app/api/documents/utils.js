export function updateMetadataNames(documents, nameMatches) {
  documents.forEach((doc) => {
    let oldMetadata = Object.assign({}, doc.metadata);
    Object.keys(doc.metadata).forEach((propertyName) => {
      if (nameMatches[propertyName]) {
        delete doc.metadata[propertyName];
        doc.metadata[nameMatches[propertyName]] = oldMetadata[propertyName];
      }
    });
  });

  return documents;
}
