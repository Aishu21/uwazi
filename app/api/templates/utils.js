import uuid from 'node-uuid';

export function generateNamesAndIds(properties) {
  return properties.map((property) => {
    property.name = property.label.replace(/[^a-z0-9]/gi, '_').toLowerCase();
    if (!property.id) {
      property.id = uuid.v4();
    }
    return property;
  });
}

export function getUpdatedNames(oldProperties = [], newProperties, prop = 'name') {
  let propertiesWithNewName = {};
  oldProperties.forEach((property) => {
    let newProperty = newProperties.find((p) => p.id === property.id);
    if (newProperty && newProperty.name !== property.name) {
      propertiesWithNewName[property[prop]] = newProperty[prop];
    }
  });

  return propertiesWithNewName;
}

export function getDeletedProperties(oldProperties = [], newProperties, prop = 'name') {
  let deletedProperties = [];

  oldProperties.forEach((property) => {
    let newProperty = newProperties.find((p) => p.id === property.id);
    if (!newProperty) {
      deletedProperties.push(property[prop]);
    }
  });

  return deletedProperties;
}
