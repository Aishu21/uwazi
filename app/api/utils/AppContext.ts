import { AsyncLocalStorage } from 'async_hooks';

interface ContextData {
  [k: string]: unknown;
}

class AppContext {
  private storage = new AsyncLocalStorage<ContextData>();

  async run(cb: () => Promise<void>, data: ContextData = {}): Promise<void> {
    return new Promise((resolve, reject) => {
      this.storage.run(data, () => {
        cb()
          .then(resolve)
          .catch(reject);
      });
    });
  }

  get(key: string) {
    const data = this.storage.getStore();
    return data && data[key];
  }

  set(key: string, value: unknown) {
    const data = this.storage.getStore();
    if (data) data[key] = value;
  }
}

const appContext = new AppContext();

export { appContext };
