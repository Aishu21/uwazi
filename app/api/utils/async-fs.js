import { promisify } from 'util';
import fs from 'fs';

export default {
  writeFile: promisify(fs.writeFile),
  exists: promisify(fs.exists),
  unlink: promisify(fs.unlink),
  rename: promisify(fs.rename),
  readFile: promisify(fs.readFile),
  readdir: promisify(fs.readdir),
};
