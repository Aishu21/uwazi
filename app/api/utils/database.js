import fetch from 'isomorphic-fetch'

let database =  {

  reset_testing_database () {
    return fetch('http://127.0.0.1:5984/uwazi_development/', {method:'DELETE'})
    .then((response) => fetch('http://127.0.0.1:5984/uwazi_development/', {method:'PUT'}))
  },

  import (fixture) {
    return fetch('http://127.0.0.1:5984/uwazi_development/_bulk_docs',
    {
      method:'POST',
      headers: { 'Accept': 'application/json', 'Content-Type': 'application/json' },
      credentials: 'same-origin',
      body: JSON.stringify(fixture)
    })
  }

}

export default database;
