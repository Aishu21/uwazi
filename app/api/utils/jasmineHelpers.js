/** @format */

const assessStatus = routeResult => {
  let status;
  const req = {};
  const next = () => {};
  const res = {
    status: code => {
      status = code;
    },
    json: () => {},
  };

  routeResult.middlewares[0](req, res, next);

  return status;
};

const assessAuthorized = (routeResult, roles = []) => {
  const nextCalledFor = [];

  roles.forEach(role => {
    const req = {
      user: { role },
      get: key => (key === 'X-Requested-With' ? 'XMLHttpRequest' : ''),
    };
    const res = { status: () => {}, json: () => {} };
    const next = () => {
      nextCalledFor.push(role);
    };
    routeResult.middlewares[0](req, res, next);
  });

  return roles.join(',') === nextCalledFor.join(',');
};

export function catchErrors(done) {
  return error => {
    if (error instanceof Error) {
      return done.fail(error.stack);
    }
    return done.fail(JSON.stringify(error));
  };
}

const matchers = {
  toNeedAuthorization() {
    return {
      compare(routeResult, expected) {
        let routeValidatesExpected = true;
        const routeCanFail = assessStatus(routeResult) === 401;

        if (expected) {
          routeValidatesExpected = assessAuthorized(routeResult, expected);
        }

        const result = { pass: routeCanFail && routeValidatesExpected };
        if (result.pass) {
          result.message = () => 'route is authorized';
        } else {
          result.message = () =>
            routeCanFail
              ? `Route is not correctly authorized for ['${expected.join("', '")}']`
              : 'Route is not authorized ! (Auth middleware should be the first one)';
        }
        return result;
      },
    };
  },
  containItems() {
    return {
      compare(actual, expected) {
        const result = {};
        const missingItems = expected.filter(item => !actual.includes(item));
        result.pass = !missingItems.length && actual.length === expected.length;
        if (result.pass) {
          result.message = 'Collections are equal';
        } else {
          result.message = `Expected: [${expected}] but got: [${actual}]`;
        }
        return result;
      },
    };
  },
  toContainEqual: jasmine.matchers ? jasmine.matchers.toContain : () => {},
};

beforeEach(() => {
  jasmine.addMatchers(matchers);
});
