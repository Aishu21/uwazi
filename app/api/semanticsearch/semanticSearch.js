import { promisify } from 'util';
import async from 'async';
import { Types } from 'mongoose';
import { search } from '../search';
import model from './model';
import resultsModel from './resultsModel';
import api from './api';
import documentsModel from '../documents';
import workers from './workerManager';
import { createError } from '../utils';

import date from 'api/utils/date.js';

export const PENDING = 'pending';
export const COMPLETED = 'completed';
export const PROCESSING = 'processing';
export const IN_PROGRESS = 'inProgress';
export const STOPPED = 'stopped';

const SEARCH_BATCH_SIZE = 5;

const eachLimitAsync = promisify(async.eachLimit);

const getSearchDocuments = async ({ documents, query }, language, user) => {
  if (documents && documents.length) {
    return documents;
  }
  const _query = { ...query, limit: 9999, searchTerm: '' };
  const res = await search.search(_query, language, user);
  return res.rows.map(document => document.sharedId);
};

const updateSearchDocumentStatus = async (searchId, sharedId, status) => model.db.findOneAndUpdate({
  _id: searchId,
  'documents.sharedId': sharedId
}, {
  $set: { 'documents.$.status': status }
}, { new: true, lean: true });

const setSearchDocumentResults = async (searchId, sharedId, results) => {
  const averageScore = results.reduce((total, curr) => total + curr.score, 0) / results.length;
  const docResults = await resultsModel.db.findOneAndUpdate({
    sharedId,
    searchId
  }, {
    sharedId,
    searchId,
    averageScore,
    results: results.sort((r1, r2) => r2.score - r1.score),
    status: COMPLETED
  }, { upsert: true, new: true });
  return docResults;
};

const removePageAnnotations = text => text.replace(/\[\[\d+\]\]/g, '');

const processDocument = async (searchId, searchTerm, sharedId, language) => {
  const [doc] = await documentsModel.get({ sharedId, language }, '+fullText');
  const { fullText } = doc;

  await updateSearchDocumentStatus(searchId, sharedId, PROCESSING);
  if (!fullText) {
    return updateSearchDocumentStatus(searchId, sharedId, COMPLETED);
  }
  Object.keys(fullText).forEach((page) => {
    fullText[page] = removePageAnnotations(fullText[page]);
  });
  const results = await api.processDocument({
    searchTerm,
    contents: fullText
  });
  const savedResults = await setSearchDocumentResults(searchId, sharedId, results);
  await updateSearchDocumentStatus(searchId, sharedId, COMPLETED);
  return savedResults;
};

const updateSearchStatus = (searchId, status) => model.save({
  _id: searchId,
  status
});

const getAllDocumentResults = async searchId => resultsModel.get({ searchId });

const processSearchLimit = async (searchId, docLimit) => {
  const searchObject = await updateSearchStatus(searchId, IN_PROGRESS);
  const { language, searchTerm } = searchObject;
  const docs = searchObject.documents
  .filter(doc => doc.status !== COMPLETED);
  const docsToSearch = docs.length > docLimit ?
    docs.slice(0, docLimit) : docs;
  await eachLimitAsync(docsToSearch, SEARCH_BATCH_SIZE, async doc => processDocument(
    searchId, searchTerm, doc.sharedId, language));
  const updatedSearch = await model.getById(searchId);
  const isNotDone = updatedSearch.documents.some(doc => doc.status !== COMPLETED);
  if (isNotDone) {
    return { updatedSearch, processedDocuments: docsToSearch.map(d => d.sharedId) };
  }
  return { updatedSearch: await updateSearchStatus(searchId, COMPLETED), processedDocuments: docsToSearch.map(d => d.sharedId) };
};

const create = async (args, language, user) => {
  const docs = await getSearchDocuments(args, language, user);
  const newSearch = {
    documents: docs.map(docId => ({
      sharedId: docId,
      status: PENDING
    })),
    status: PENDING,
    searchTerm: args.searchTerm,
    query: args.query,
    language,
    creationDate: date.currentUTC()
  };
  const savedSearch = await model.save(newSearch);
  workers.notifyNewSearch(savedSearch._id);
  return savedSearch;
};

const filterSearchResults = async (searchId, { skip = 0, limit = 50, threshold = 0.3, minRelevantSentences = 5 }) =>
  resultsModel.db.aggregate([
    {
      $match: { searchId: Types.ObjectId(searchId) }
    },
    {
      $project: {
        searchId: 1,
        sharedId: 1,
        status: 1,
        totalResults: { $size: '$results' },
        results: { $filter: { input: '$results', as: 'result', cond: { $gte: ['$$result.score', threshold] } } }
      }
    },
    {
      $project: {
        totalResults: 1,
        searchId: 1,
        sharedId: 1,
        status: 1,
        results: 1,
        numRelevant: { $size: '$results' },
        percentage: { $divide: [{ $size: '$results' }, '$totalResults'] }
      }
    },
    {
      $match: { numRelevant: { $gte: minRelevantSentences } }
    },
    {
      $sort: { percentage: -1 }
    },
    {
      $skip: skip
    },
    {
      $limit: limit
    }
  ]);

const getSearch = async (searchId) => {
  const theSearch = await model.getById(searchId);
  if (!theSearch) {
    throw createError('Search not found', 404);
  }
  const results = await filterSearchResults(searchId, {});
  const docIds = results.map(r => r.sharedId);
  const docs = await documentsModel.get({ sharedId: { $in: docIds }, language: theSearch.language });
  const docsWithResults = docs.map(doc => (
    {
      ...doc,
      semanticSearch: results.find(res => res.sharedId === doc.sharedId)
    }
  ));
  theSearch.results = docsWithResults;
  return theSearch;
};

const getDocumentResultsByIds = async (searchId, docIds) => {
  const theSearch = searchId._id ? searchId : await model.getById(searchId);
  const results = await resultsModel.get({ searchId, sharedId: { $in: docIds } });
  const docs = await documentsModel.get({ sharedId: { $in: docIds }, language: theSearch.language });
  const docsWithResults = docs.map(doc => (
    {
      ...doc,
      semanticSearch: results.find(res => res.sharedId === doc.sharedId)
    }
  ));
  return docsWithResults;
};

const getSearchesByDocument = async (docId) => {
  const results = await resultsModel.get({ sharedId: docId });
  const searchPromises = results.map(({ searchId }) => model.getById(searchId));
  const searches = await Promise.all(searchPromises);
  searches.map((theSearch, index) => ({
    ...theSearch,
    results: results[index]
  }));
  return searches;
};

const deleteSearch = async (searchId) => {
  const res = await model.delete(searchId);
  if (res.n !== 1) {
    throw createError('Search not found', 404);
  }
  await resultsModel.delete({ searchId });
  return { deleted: true };
};

const stopSearch = async (searchId) => {
  const res = await model.db.updateOne({
    _id: searchId,
    status: { $in: [IN_PROGRESS, PENDING] }
  }, {
    $set: { status: STOPPED }
  });
  if (!res.n) {
    throw createError('Only running searches can be stopped', 400);
  }
  return model.getById(searchId);
};

const resumeSearch = async (searchId) => {
  const res = await model.db.updateOne({
    _id: searchId,
    status: { $in: [STOPPED] }
  }, {
    $set: { status: PENDING }
  });
  if (!res.n) {
    throw createError('Only stopped searches can be resumed', 400);
  }
  workers.notifyNewSearch(searchId);
  return model.getById(searchId);
};

const getAllSearches = () => model.get().sort('-creationDate');
const getInProgress = async () => model.get({ status: IN_PROGRESS }).sort('-creationDate');
const getPending = async () => model.get({ status: PENDING }).sort('-creationDate');

const semanticSearch = {
  _id: Math.random(),
  create,
  processDocument,
  processSearchLimit,
  getAllDocumentResults,
  getDocumentResultsByIds,
  getAllSearches,
  getPending,
  getInProgress,
  getSearch,
  getSearchesByDocument,
  deleteSearch,
  stopSearch,
  resumeSearch
};

export default semanticSearch;
