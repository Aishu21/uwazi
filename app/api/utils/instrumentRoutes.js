import * as utils from 'api/utils';

const executeRoute = (method, routePath, req = {}, res, next = () => {}, app, runRoute = true) => {
  const args = app[method].calls.allArgs().find(a => a[0] === routePath);
  if (!args) {
    throw new Error(`Route ${method.toUpperCase()} ${routePath} is not defined`);
  }
  const result = new Promise((resolve, reject) => {
    let statusCode;

    res.status = (code) => {
      statusCode = code;
    };

    res.download = jasmine.createSpy('download').and.callFake((response) => {
      resolve(response);
    });

    res.json = jasmine.createSpy('json').and.callFake((response) => {
      if (statusCode) {
        response.status = statusCode;
      }
      resolve(response);
    });

    if (!args[1]) {
      return reject(new Error('route function has not been defined !'));
    }

    if (!runRoute) {
      return resolve();
    }

    const mockedNext = (error) => {
      next(error);
      resolve();
    };

    return args[args.length - 1](req, res, mockedNext);
  });

  if (args) {
    result.middlewares = args.slice(1, -1).filter(arg => !arg.isJoi);
    const validation = args.slice(1, -1).find(arg => arg.isJoi);
    if (validation) {
      result.validation = validation.describe();
    }
  }

  return result;
};

export default (route, io) => {
  const app = jasmine.createSpyObj('app', ['get', 'post', 'delete', 'use']);
  const originalValidateRequest = utils.validateRequest;
  spyOn(utils, 'validateRequest').and.callFake(schema => schema);
  route(app, io);
  utils.validateRequest = originalValidateRequest;

  const get = (routePath, req, res = {}, next) => executeRoute('get', routePath, req, res, next, app);
  get.validation = (routePath, req, res = {}, next) => executeRoute('get', routePath, req, res, next, app, false).validation;

  const post = (routePath, req, res = {}, next) => executeRoute('post', routePath, req, res, next, app);
  post.validation = (routePath, req, res = {}, next) => executeRoute('post', routePath, req, res, next, app, false).validation;

  const remove = (routePath, req, res = {}, next) => executeRoute('delete', routePath, req, res, next, app);
  remove.validation = (routePath, req, res = {}, next) => executeRoute('delete', routePath, req, res, next, app, false).validation;

  const instrumentedRoute = {
    get,
    delete: remove,
    post,
  };

  return instrumentedRoute;
};
