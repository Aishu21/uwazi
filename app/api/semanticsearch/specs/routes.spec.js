import semanticSearch from '../semanticSearch';
import { catchErrors } from 'api/utils/jasmineHelpers';
import semanticRoutes from '../routes.js';
import instrumentRoutes from '../../utils/instrumentRoutes';

describe('search routes', () => {
  let routes;
  beforeEach(() => {
    routes = instrumentRoutes(semanticRoutes);
  });

  describe('POST /api/semantic-search', () => {
    it('should have a validation schema', () => {
      expect(routes.post.validation('/api/semantic-search')).toMatchSnapshot();
    });

    it('should create a semantic search', (done) => {
      const result = { status: 'pending' };
      const body = {
        searchTerm: 'term',
        query: {},
        documents: []
      };
      jest.spyOn(semanticSearch, 'create').mockResolvedValue(result);
      const req = { language: 'en', user: 'user', body };

      routes.post('/api/semantic-search', req)
      .then((res) => {
        expect(res).toEqual(result);
        expect(semanticSearch.create).toHaveBeenCalledWith(body, 'en', 'user');
        done();
      })
      .catch(catchErrors(done));
    });
  });

  describe('GET /api/semantic-search', () => {
    it('should return all searches', (done) => {
      const result = [{ _id: 's1' }, { _id: 's2' }];
      jest.spyOn(semanticSearch, 'getAllSearches').mockResolvedValue(result);
      const req = {};

      routes.get('/api/semantic-search', req)
      .then((response) => {
        expect(response).toEqual(result);
        expect(semanticSearch.getAllSearches).toHaveBeenCalled();
        done();
      })
      .catch(catchErrors(done));
    });
  });

  describe('GET /api/semantic-search/:searchId', () => {
    it('should return specified search', (done) => {
      const result = { _id: 's1' };
      jest.spyOn(semanticSearch, 'getSearch').mockResolvedValue(result);
      const req = { params: { searchId: 's1' } };

      routes.get('/api/semantic-search/:searchId', req)
      .then((response) => {
        expect(response).toEqual(result);
        expect(semanticSearch.getSearch).toHaveBeenCalledWith('s1');
        done();
      })
      .catch(catchErrors(done));
    });
  });

  describe('DELETE /api/semantic-search/:searchId', () => {
    it('should delete specified', (done) => {
      const result = { _id: 's1' };
      jest.spyOn(semanticSearch, 'deleteSearch').mockResolvedValue(result);
      const req = { params: { searchId: 's1' } };

      routes.delete('/api/semantic-search/:searchId', req)
      .then((response) => {
        expect(response).toEqual(result);
        expect(semanticSearch.deleteSearch).toHaveBeenCalledWith('s1');
        done();
      })
      .catch(catchErrors(done));
    });
  });

  describe('POST /api/semantic-search/:searchId/stop', () => {
    it('should stop the search', (done) => {
      const result = { _id: 's1' };
      jest.spyOn(semanticSearch, 'stopSearch').mockResolvedValue(result);
      const req = { params: { searchId: 's1' } };

      routes.post('/api/semantic-search/:searchId/stop', req)
      .then((response) => {
        expect(response).toEqual(result);
        expect(semanticSearch.stopSearch).toHaveBeenCalledWith('s1');
        done();
      })
      .catch(catchErrors(done));
    });
  });

  describe('POST /api/semantic-search/:searchId/resume', () => {
    it('should resume the search', (done) => {
      const result = { _id: 's1' };
      jest.spyOn(semanticSearch, 'resumeSearch').mockResolvedValue(result);
      const req = { params: { searchId: 's1' } };

      routes.post('/api/semantic-search/:searchId/resume', req)
      .then((response) => {
        expect(response).toEqual(result);
        expect(semanticSearch.resumeSearch).toHaveBeenCalledWith('s1');
        done();
      })
      .catch(catchErrors(done));
    });
  });
});
