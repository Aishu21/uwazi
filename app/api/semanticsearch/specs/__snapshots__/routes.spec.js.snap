// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search routes GET /api/semantic-search should have a validation schema 1`] = `
Object {
  "children": Object {
    "limit": Object {
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "rules": Array [
        Object {
          "arg": 0,
          "name": "min",
        },
      ],
      "type": "number",
    },
    "minRelevantSentences": Object {
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "rules": Array [
        Object {
          "arg": 0,
          "name": "min",
        },
      ],
      "type": "number",
    },
    "searchId": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "skip": Object {
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "rules": Array [
        Object {
          "arg": 0,
          "name": "min",
        },
      ],
      "type": "number",
    },
    "threshold": Object {
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "rules": Array [
        Object {
          "arg": 0,
          "name": "min",
        },
      ],
      "type": "number",
    },
  },
  "type": "object",
}
`;

exports[`search routes GET /api/semantic-search/list should have a validation schema 1`] = `
Object {
  "children": Object {
    "minRelevantSentences": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "rules": Array [
        Object {
          "arg": 0,
          "name": "min",
        },
      ],
      "type": "number",
    },
    "searchId": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "threshold": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "rules": Array [
        Object {
          "arg": 0,
          "name": "min",
        },
      ],
      "type": "number",
    },
  },
  "type": "object",
}
`;

exports[`search routes POST /api/semantic-search should have a validation schema 1`] = `
Object {
  "children": Object {
    "documents": Object {
      "flags": Object {
        "sparse": false,
      },
      "items": Array [
        Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
      ],
      "type": "array",
    },
    "query": Object {
      "type": "object",
    },
    "searchTerm": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;
